valast.go

```diff
@@ -191,7 +191,16 @@ func basicLit(vv reflect.Value, kind token.Token, builtinType string, v interfac
 		return Result{}, err
 	}
 	if opt.Unqualify && vv.Type().Name() == builtinType && vv.Type().PkgPath() == "" {
```

Comment:
Thank you for the PR! Unfortunately I don't think this is the right fix in the current form.

In this `if` case, `opt.Unqualify` signals:

```
	// Unqualify, if true, indicates that types should be unqualified. e.g.:
	//
	// 	int(8)           -> 8
	// 	Bar{}            -> Bar{}
	// 	string("foobar") -> "foobar"
	//
	// This is set to true automatically when operating within a context where type qualification
	// is definitively not needed, e.g. when producing values for a struct or map.
	Unqualify bool
```

That is, if `opt.Unqualify` is `true` (this if case is executing) then the value should not have a qualified type.

In the case of #15, it seems like the issue is that we are producing:

```
valast.str{ExpirationSeconds: valast.Addr(3607).(*int64)}
```

instead of the expected:

```
valast.str{ExpirationSeconds: valast.Addr(int64(3607)).(*int64)}
```

I think that what this means is that we're not correctly setting `Unqualify` to `false` (so the `int64()` is added) when we emit the expression for `valast.Addr(expr)` - the fix will need to be located somewhere around that code.

We emit `valast.Addr` here: https://github.com/hexops/valast/blob/c24911aa506e196bc226fd72cbc4a04ac0dc9b59/valast.go#L488-L501

The expression is here: https://github.com/hexops/valast/blob/c24911aa506e196bc226fd72cbc4a04ac0dc9b59/valast.go#L495

I think maybe we just need to move [this code](https://github.com/hexops/valast/blob/c24911aa506e196bc226fd72cbc4a04ac0dc9b59/valast.go#L480-L484) into [this conditional](https://github.com/hexops/valast/blob/c24911aa506e196bc226fd72cbc4a04ac0dc9b59/valast.go#L486) and set `opt.Unqualify = false` before invoking it - so that `int64()` gets added.

Reply:
Thank you for the detailed response. The idea was to write some code to spur this kind of feedback. You did not disappoint! :1st_place_medal: 

I'll give this a shot tomorrow and resubmit.

Reply:
More information on this, the following code results in a build failure:

```go
	type str struct {
		ExpirationSeconds *int32
	}

	broken := str{ExpirationSeconds: valast.Addr(3607).(*int32)}
	fmt.Println(broken)
```

In short, Unqualified in a struct yields code that is invalid.

```go
panic: interface conversion: interface {} is *int, not *int32
```

What are your thoughts on always forcing qualified types when working with pointers?

Reply:
I would expect your snippet above to be:

```diff
	type str struct {
		ExpirationSeconds *int32
	}

-	broken := str{ExpirationSeconds: valast.Addr(3607).(*int32)}
+	broken := str{ExpirationSeconds: valast.Addr(int32(3607)).(*int32)}
	fmt.Println(broken)
```

Then you would not get such a panic, as the type would be qualified. This is only needed when the input value is a integer literal and the type assertion is not `int`.

I'm happy to look into fixing this for you if you want to convert this into a bug report BTW

Reply:
A bug exists for this issue at #15.

I would greatly appreciate any help here. I'm not partial/biased to this code or the time I put in, so if you have a solution in mind, I will not feel bad if you end up closing this and replacing it with your own implementation. :)

I tried a workaround by setting Unqualified to false, but it still emits unqualified code. I'll open another bug for that soon.